README.md
# Bad Apple!! player

Want Bad Apple!! in VS Code? No? Here it is anyway...

NOTE: For best results, use monospace.

## About

![Screenshot](images/screenshot.png)
Theme: [Bearded Theme HC Ebony](https://github.com/BeardedBear/bearded-theme)

This extension comes with these commands accessible by the command palette (default is `Ctrl+Shift+P`):
* `Bad Apple!!: Play` - Play Bad Apple!! in your active text editor.
* `Bad Apple!!: Pad` - Pads text so Bad Apple!! can be watched in full.
* `Bad Apple!!: Kill` - Stops all Bad Apple!! instances.

## Extension Settings

This extension contributes the following settings:

* `bad-apple-vscode-player.targetFrames`: Target FPS to display Bad Apple!!
* `bad-apple-vscode-player.opacity`: Opacity of Bad Apple!!
* `bad-apple-vscode-player.textOpacity`: Opacity of text when running Bad Apple!!
* `bad-apple-vscode-player.switchEditor`: Whether to switch to the current active editor.

## Known Issues

Currently, this extension is unable to resize for different dimensions.

This extension also works poorly without a monospace font.

---

**Enjoy!**


CHANGELOG.md
# Change Log

All notable changes to the "bad-apple-vscode-player" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## 0.0.2

- Documented commands in README

## 0.0.1

- Initial release


gen/gen.py
import cv2
import numpy
import json

# The video is sourced from https://archive.org/details/TouhouBadApple

def get_frame(video: cv2.VideoCapture, frame_id: int) -> numpy.ndarray:
    video.set(cv2.CAP_PROP_POS_FRAMES, frame_id)
    status, frame = video.read()
    assert status, f"Error at: {frame_id}"
    return frame.mean(2)

def resize_frame(frame: numpy.ndarray, width: int, height: int) -> numpy.ndarray:
    y, x = frame.shape
    cropped_frame = frame[:y // height * height, :x // width * width]
    new_frame = cropped_frame.reshape(height, y // height, width, x // width)
    return new_frame.mean((1, 3))

def encode(frame: numpy.ndarray) -> list:
    new_frame = []
    initial = [0]
    final = [len(frame[0])]
    for row in frame:
        new_row = row[:-1] != row[1:]
        new_frame.append([int(row[0]), initial + [int(i) for i in new_row.nonzero()[0] + 1] + final])
    return new_frame

def write(data: dict) -> None:
    with open("data.json", "w+") as f:
        json.dump(data, f)

def main() -> None:
    video = cv2.VideoCapture("./gen/Touhou - Bad Apple.mp4")
    length = int(video.get(cv2.CAP_PROP_FRAME_COUNT))

    frames = []

    for i in range(length):
        frame = get_frame(video, i)
        frame = resize_frame(frame, 80, 30) > 128
        frame = encode(frame)
        frames.append(frame)

    data = {}
    data["n"] = length
    data["frames"] = frames

    write(data)

main()


src/extension.ts
import * as vscode from "vscode";
const data: {[key: string]: any} = require("./../data.json");

function generateDecorations(): [vscode.TextEditorDecorationType, vscode.TextEditorDecorationType] {
	let opacity = vscode.workspace.getConfiguration("bad-apple-vscode-player").get("opacity", 1);
	let textOpacity = vscode.workspace.getConfiguration("bad-apple-vscode-player").get("textOpacity", 1);

	let dark = vscode.window.createTextEditorDecorationType({backgroundColor: `rgba(0, 0, 0, ${opacity})`, opacity: `${textOpacity}`});
	let light = vscode.window.createTextEditorDecorationType({backgroundColor: `rgba(255, 255, 255, ${opacity})`, opacity: `${textOpacity}`});

	return [dark, light];
}

function generateFrame(i: number, offset: number) {
	let frame = data.frames[i];
	let newFrame: [vscode.Range[], vscode.Range[]] = [[], []];
	for (let row = 0; row < frame.length; row++) {
		for (let j = 0; j < frame[row][1].length - 1; j++){
			newFrame[(j & 1) ^ frame[row][0]].push(new vscode.Range(
				row + offset, frame[row][1][j],
				row + offset, frame[row][1][j + 1]
			));
		}
	}
	return newFrame;
}

const sleep = (ms: number) => new Promise(r => setTimeout(r, ms));

async function frameSync(fps: number) {
	let frameTime = 1000/fps;
	let time = new Date();

	await sleep(frameTime - time.getMilliseconds() % frameTime);
}

let killSignal = false;

export function activate(context: vscode.ExtensionContext) {
	let pad = vscode.commands.registerTextEditorCommand("bad-apple-vscode-player.pad", async function () {
		let editor = vscode.window.activeTextEditor;
		if (editor === undefined) {return;}

		let edits: [vscode.Position, string][] = [];
		let lines = editor.document.lineCount;

		for (let n = 0; n < lines; n++) {
			let line = editor.document.lineAt(n);
			if (line.text.length >= 80) {continue;}
			let text = " ".repeat(80 - line.text.length);
			edits.push([line.range.end, text]);
		}

		if (lines < 30) {
			edits[lines - 1][1] += Array(30 - lines).fill("\n" + " ".repeat(80)).join("");
		}
		editor.edit(editBuilder => {edits.map(x => editBuilder.insert(...x));});
	});

	context.subscriptions.push(pad);

	let play = vscode.commands.registerCommand("bad-apple-vscode-player.play", async function () {
		killSignal = false;
		let fps = vscode.workspace.getConfiguration("bad-apple-vscode-player").get("targetFrames", 15);
		let switchEditor = vscode.workspace.getConfiguration("bad-apple-vscode-player").get("switchEditor", false);
		let decorations = generateDecorations();
		let editor = vscode.window.activeTextEditor;

		let editorHandler = switchEditor ? vscode.window.onDidChangeActiveTextEditor((e) => {
			editor?.setDecorations(decorations[0], []);
			editor?.setDecorations(decorations[1], []);
			editor = e;
			offset = e?.visibleRanges[0].start.line || 0;
		}) : undefined;

		if (editorHandler === undefined && switchEditor === undefined) {
			vscode.window.showErrorMessage("Unable to use active text editor.");
			return; // Early return to avoid computation
		}

		let offset = editor === undefined ? 0 : editor.visibleRanges[0].start.line;
		let lineHandler = vscode.window.onDidChangeTextEditorVisibleRanges((e) => {
			if (e.textEditor === editor) {offset = e.visibleRanges[0].start.line;}
		});

		for (let i = 0; i < data["n"]; i += 30/fps) {
			let frame = generateFrame(Math.round(i), offset);
			editor?.setDecorations(decorations[0], frame[0]);
			editor?.setDecorations(decorations[1], frame[1]);
			await frameSync(fps);
			if (killSignal) {break;}
		}

		editor?.setDecorations(decorations[0], []);
		editor?.setDecorations(decorations[1], []);

		lineHandler.dispose();
		if (switchEditor) {editorHandler?.dispose();}
	});

	context.subscriptions.push(play);

	let kill = vscode.commands.registerCommand("bad-apple-vscode-player.kill", async function () {
		killSignal = true;
	});

	context.subscriptions.push(kill);
}

export function deactivate() {}
